File: zig/1545.zig
const std = @import("std");

fn solution(allocator: std.mem.Allocator, n: usize, k: usize) !u8 {
    var table = try allocator.alloc([]u8, n);
    table[0] = try allocator.alloc(u8, 1);
    table[0][0] = '0';

    defer {
        for (0..n) |idx| {
            allocator.free(table[idx]);
        }

        allocator.free(table);
    }

    for (1..n) |idx| {
        const last = table[idx - 1];
        var newString: std.ArrayList(u8) = .empty;
        defer newString.deinit(allocator);

        for (last) |ch| {
            try newString.append(allocator, ch);
        }

        try newString.append(allocator, '1');

        // invert
        var inverted: std.ArrayList(u8) = .empty;
        defer inverted.deinit(allocator);

        for (last) |ch| {
            if (ch == '1') {
                try inverted.append(allocator, '0');
            } else {
                try inverted.append(allocator, '1');
            }
        }

        // reverse
        for (0..inverted.items.len / 2) |i| {
            const temp = inverted.items[i];
            inverted.items[i] = inverted.items[inverted.items.len - 1 - i];
            inverted.items[inverted.items.len - 1 - i] = temp;
        }

        for (inverted.items) |ch| {
            try newString.append(allocator, ch);
        }

        table[idx] = try newString.toOwnedSlice(allocator);
    }

    return table[n - 1][k - 1];
}

const expect = std.testing.expect;

test "test case #1" {
    const allocator = std.testing.allocator;

    const n = 3;
    const k = 1;
    const expected = '0';

    const sol = try solution(allocator, n, k);

    try expect(sol == expected);
}

test "test case #2" {
    const allocator = std.testing.allocator;

    const n = 4;
    const k = 11;
    const expected = '1';

    const sol = try solution(allocator, n, k);

    try expect(sol == expected);
}

File: go/1545.go
package main

import (
	"strings"
)

func invert(s string) string {
	var sb strings.Builder

	for _, ch := range s {
		if ch == '0' {
			sb.WriteRune('1')
		} else {
			sb.WriteRune('0')
		}
	}

	return sb.String()
}

func reverse(s string) string {
	var sb strings.Builder
	idx := len(s) - 1

	for idx >= 0 {
		sb.WriteByte(s[idx])
		idx--
	}

	return sb.String()
}

func findKthBit(n int, k int) byte {
	table := []string{"0"}

	for len(table) < n {
		var sb strings.Builder

		last := table[len(table)-1]
		sb.WriteString(last)
		sb.WriteString("1")
		sb.WriteString(reverse(invert(last)))

		table = append(table, sb.String())
	}

	return table[len(table)-1][k-1]
}
